src/msg/producer/writer/consumer_writer.go

```diff
@@ -278,69 +337,109 @@ func (w *consumerWriterImpl) continueFn(int) bool {
 	return !w.isClosed()
 }
 
-func (w *consumerWriterImpl) readAcks() error {
-	if !w.validConn.Load() {
+func (w *consumerWriterImpl) readAcks(idx int) error {
+	w.writeState.Lock()
+	validConns := w.writeState.validConns
+	conn := w.writeState.conns[idx]
+	w.writeState.Unlock()
+	if !validConns {
 		w.m.readInvalidConn.Inc(1)
 		return errInvalidConnection
 	}
+
+	// Read from decoder, safe to read from acquired decoder as not re-used.
 	// NB(cw) The proto needs to be cleaned up because the gogo protobuf
 	// unmarshalling will append to the underlying slice.
-	w.ack.Metadata = w.ack.Metadata[:0]
-	w.decodeLock.Lock()
-	err := w.decoder.Decode(&w.ack)
-	w.decodeLock.Unlock()
+	conn.ack.Metadata = conn.ack.Metadata[:0]
+	err := conn.decoder.Decode(&conn.ack)
 	if err != nil {
-		w.notifyReset()
+		w.notifyReset(err)
 		w.m.decodeError.Inc(1)
 		return err
 	}
-	for _, m := range w.ack.Metadata {
+	for _, m := range conn.ack.Metadata {
 		if err := w.router.Ack(newMetadataFromProto(m)); err != nil {
 			w.m.ackError.Inc(1)
 			// This is fine, usually this means the ack has been acked.
 			w.logger.Error("could not ack metadata", zap.Error(err))
 		}
 	}
+
 	return nil
 }
 
 func (w *consumerWriterImpl) Close() {
-	if !w.closed.CAS(false, true) {
+	w.writeState.Lock()
+	wasClosed := w.writeState.closed
+	w.writeState.closed = true
+	w.writeState.Unlock()
+
+	if wasClosed {
 		return
 	}
+
 	close(w.doneCh)
+
 	w.wg.Wait()
 }
 
-func (w *consumerWriterImpl) notifyReset() {
+func (w *consumerWriterImpl) notifyReset(err error) {
 	select {
 	case w.resetCh <- struct{}{}:
+		if err != nil {
+			w.logger.Error("connection error", zap.Error(err))
+		}
 	default:
 	}
 }
 
 func (w *consumerWriterImpl) isClosed() bool {
-	return w.closed.Load()
+	w.writeState.Lock()
+	defer w.writeState.Unlock()
+	return w.writeState.closed
 }
 
-func (w *consumerWriterImpl) reset(conn io.ReadWriteCloser) {
-	// Close the connection to wake up potential blocking encode/decode calls.
-	w.conn.Close()
+type resetOptions struct {
+	connections []io.ReadWriteCloser
+	at          time.Time
+	validConns  bool
+}
+
+func (w *consumerWriterImpl) reset(opts resetOptions) {
+	w.writeState.Lock()
+	prevConns := w.writeState.conns
+	defer func() {
```

Comment:
Would this work better replacing the conns in place? Or it the size of conns not guaranteed to be the same every reset?